extensions/amp-analytics/0.1/events.js

```diff
@@ -447,24 +722,128 @@ export class TimerEventTracker extends EventTracker {
+        'Bad timer start specification');
+    const timerStop = 'stopSpec' in timerSpec ? timerSpec['stopSpec'] : null;
+    user().assert((!timerStart && !timerStop) || typeof timerStop == 'object',
+        'Bad timer stop specification');
+
+    const timerId = this.generateTimerId_();
+    const isUnstoppableTimer = !timerStop;
+    let startBuilder;
+    let stopBuilder;
+    if (!!timerStart) {
+      const startTracker = this.getTracker(timerStart);
+      user().assert(startTracker, 'Cannot track timer start');
+      startBuilder = startTracker.add.bind(startTracker, context,
+          timerStart['on'], timerStart,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
+    }
+    if (!isUnstoppableTimer) {
+      const stopTracker = this.getTracker(timerStop);
+      user().assert(stopTracker, 'Cannot tracker timer stop');
+      stopBuilder = stopTracker.add.bind(stopTracker, context,
+          timerStop['on'], timerStop,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
+    }
 
-    const win = this.root.ampdoc.win;
-    const intervalId = win.setInterval(() => {
-      listener(this.createEvent_(eventType));
-    }, interval * 1000);
-    this.trackers_.push(intervalId);
-    win.setTimeout(() => {
-      this.removeTracker_(intervalId);
-    }, maxTimerLength * 1000);
-    if (callImmediate) {
-      listener(this.createEvent_(eventType));
+    const timerHandler = new TimerEventHandler(interval, maxTimerLength,
+        isUnstoppableTimer, callImmediate, startBuilder, stopBuilder);
+    this.trackers_[timerId] = timerHandler;
+
+    if (!timerStart) {
+      // Timer starts on load.
+      this.startTimer_(timerId, eventType, listener);
+    } else {
+      // Timer starts on event.
+      timerHandler.listenForStart();
     }
     return () => {
-      this.removeTracker_(intervalId);
+      this.removeTracker_(timerId);
     };
   }
 
   /**
+   * @return {number}
+   * @private
+   */
+  generateTimerId_() {
+    return ++this.timerIdSequence_;
+  }
+
+  /**
+   * @param {!JsonObject} config
+   * @return {?EventTracker}
+   */
+  getTracker(config) {
+    const eventType = user().assertString(config['on']);
+    const trackerKey = getTrackerKeyName(eventType);
+
+    return this.root.getTrackerForWhitelist(
+        trackerKey, getTrackerTypesForTimerEventTracker());
+  }
+
+  /**
+   * Toggles which listeners are active depending on timer state, so no race
+   * conditions can occur in the case where the timer starts and stops on the
+   * same event type from the same target.
+   * @param {number} timerId
+   * @param {string} eventType
+   * @param {function(!AnalyticsEvent)} listener
+   * @private
+   */
+  handleTimerToggle_(timerId, eventType, listener) {
+    const timerHandler = this.trackers_[timerId];
+    if (timerHandler.isRunning()) {
+      // Stop timer and listen for start.
+      this.stopTimer_(timerId);
+      if (timerHandler.canListenForStart()) {
```

Comment:
This is kind of related to my last comment. Is it possible that we go ahead and call `timerHandler.listenForStart()` directly on timer stopped. And then `timerHandler` can take care of `canListenForStart` and decide whether to start listening again or not. 

I think I am trying to separate the concept of `timerEventTracker` and `timerEventHandler` more strictly. Can we let `TimerEventTracker` handles sending timer event correctly, while `TimerEventHandler` handles only the start and stop point. 

I am thinking of something like 
```
timerEventTracker {
  startTimer_() {
    setTimeInterval....
  }

  stopTimer_() {
    clearTimeInterval...
  }

  new timerEventHandler(..., ..., startTimer_, stopTimer_);
}
```
Then 1timerEventTracker` knows nothing about timerStart/timerStop. All it needs to do is to start timerInterval on startTime_ call, and clear timerInterval on stopTimer_ and final clean up. 

WDYT?

Reply:
Done.